src/expr/src/scalar/func/impls/timestamp.rs

```diff
@@ -61,38 +61,182 @@ sqlfunc!(
 sqlfunc!(
     #[sqlname = "timestamp_with_time_zone_to_date"]
     #[preserves_uniqueness = false]
-    #[inverse = to_unary!(super::CastDateToTimestampTz)]
+    #[inverse = to_unary!(super::CastDateToTimestampTz(None))]
     #[is_monotone = true]
     fn cast_timestamp_tz_to_date(a: CheckedTimestamp<DateTime<Utc>>) -> Result<Date, EvalError> {
         Ok(a.naive_utc().date().try_into()?)
     }
 );
 
-sqlfunc!(
-    #[sqlname = "timestamp_to_timestamp_with_time_zone"]
-    #[preserves_uniqueness = true]
-    #[inverse = to_unary!(super::CastTimestampTzToTimestamp)]
-    #[is_monotone = true]
-    fn cast_timestamp_to_timestamp_tz(
+#[derive(
+    Arbitrary, Ord, PartialOrd, Clone, Debug, Eq, PartialEq, Serialize, Deserialize, Hash, MzReflect,
+)]
+pub struct CastTimestampToTimestampTz(pub Option<TimestampPrecision>);
+
+impl<'a> EagerUnaryFunc<'a> for CastTimestampToTimestampTz {
+    type Input = CheckedTimestamp<NaiveDateTime>;
+    type Output = Result<CheckedTimestamp<DateTime<Utc>>, EvalError>;
+
+    fn call(
+        &self,
         a: CheckedTimestamp<NaiveDateTime>,
     ) -> Result<CheckedTimestamp<DateTime<Utc>>, EvalError> {
-        DateTime::<Utc>::from_utc(a.into(), Utc)
-            .try_into()
-            .err_into()
+        let mut out = CheckedTimestamp::try_from(DateTime::<Utc>::from_utc(a.into(), Utc))?;
+        out.round_to_precision(self.0);
+        Ok(out)
     }
-);
 
-sqlfunc!(
-    #[sqlname = "timestamp_with_time_zone_to_timestamp"]
-    #[preserves_uniqueness = true]
-    #[inverse = to_unary!(super::CastTimestampToTimestampTz)]
-    #[is_monotone = true]
-    fn cast_timestamp_tz_to_timestamp(
+    fn output_type(&self, input: ColumnType) -> ColumnType {
+        ScalarType::TimestampTz { precision: self.0 }.nullable(input.nullable)
+    }
+
+    fn preserves_uniqueness(&self) -> bool {
+        // Only preserves uniqueness if the output type is as precise as the
```

Comment:
> but we're not provided the input type here.

This note makes it sound like it can't be provided the input type, rather than stating we just need to do a bit of extra typing to make that happen.

Same comment applies to all functions that include this note.

Reply:
OOC how would the extra typing work? I see this method is part of the `EagerUnaryFunc` trait. The input is only provided to the `call` function, but not here.

Reply:
In [`typeconv.rs`](https://github.com/MaterializeInc/materialize/pull/21570/files#diff-b945a38f3da415078dc2535b510ccea6e3d75cd0b441e34426418a3543cb702bR374), we would need to get the precision of the `from_type` as we currently do for the `to_type`.

Reply:
@sploiselle Ah you mean creating a separate type when the precision is same (I was incorrectly thinking how to pass the information to `preserves_uniqueness` method). I just added a TODO, will address in the follow up PR with the renames.

Reply:
Updated!